raiden_contracts/tests/test_channel_unlock.py

```diff
@@ -6,12 +6,203 @@
+    # Mock pending transfers data for a reopened channel
+    pending_transfers_tree_2 = get_pending_transfers_tree(web3, [1, 3, 5], [2, 4], settle_timeout)
+    reveal_secrets(A, pending_transfers_tree_2.unlockable)
+    # Settle the channel again
+    create_settled_channel(
+        A,
+        pending_transfers_tree_2.locked_amount,
+        pending_transfers_tree_2.merkle_root,
+        B,
+        0,
+        EMPTY_MERKLE_ROOT,
+        settle_timeout,
+    )
+    # 2nd unlocks should go through
+    token_network.functions.unlock(
+        A,
+        B,
+        pending_transfers_tree_2.packed_transfers,
+    ).transact({'from': A})
+
+    # Edge channel life-cycle: open -> settle -> open -> settle ->  unlock1 -> unlock2
+
+    # Mock pending transfers data
+    pending_transfers_tree_1 = get_pending_transfers_tree(web3, [1, 3, 5], [2, 4], settle_timeout)
+    reveal_secrets(A, pending_transfers_tree_1.unlockable)
+
+    # Settle the channel
+    create_settled_channel(
+        A,
+        pending_transfers_tree_1.locked_amount,
+        pending_transfers_tree_1.merkle_root,
+        B,
+        0,
+        EMPTY_MERKLE_ROOT,
+        settle_timeout,
+    )
+    # Mock pending transfers data for a reopened channel
+    pending_transfers_tree_2 = get_pending_transfers_tree(web3, [1, 3, 5], [2, 4], settle_timeout)
+    reveal_secrets(A, pending_transfers_tree_2.unlockable)
+    # Settle the channel again
+    create_settled_channel(
+        A,
+        pending_transfers_tree_2.locked_amount,
+        pending_transfers_tree_2.merkle_root,
+        B,
+        0,
+        EMPTY_MERKLE_ROOT,
+        settle_timeout,
+    )
+    # Both old and new unlocks should go through
+    token_network.functions.unlock(
+        A,
+        B,
+        pending_transfers_tree_2.packed_transfers,
+    ).transact({'from': A})
+    token_network.functions.unlock(
+        A,
+        B,
+        pending_transfers_tree_1.packed_transfers,
+    ).transact({'from': A})
+
+
+def test_channel_unlock_both_participants(
+        web3,
+        custom_token,
+        token_network,
+        secret_registry_contract,
+        create_channel,
+        channel_deposit,
+        get_accounts,
+        close_and_update_channel,
+):
+    (A, B) = get_accounts(2)
+    settle_timeout = 8
+
+    values_A = ChannelValues(
+        deposit=100,
+        transferred=5,
+    )
+    values_B = ChannelValues(
+        deposit=100,
+        transferred=40,
+    )
+
+    # Create channel and deposit
+    create_channel(A, B, settle_timeout)[0]
+    channel_deposit(A, values_A.deposit, B)
+    channel_deposit(B, values_B.deposit, A)
+
+    # Mock pending transfers data for A
+    pending_transfers_tree_A = get_pending_transfers_tree(web3, [1, 3, 5], [2, 4], settle_timeout)
+    locksroot_bytes = get_merkle_root(pending_transfers_tree_A.merkle_tree)
+    values_B.locksroot = '0x' + locksroot_bytes.hex()
+    values_B.locked = get_locked_amount(pending_transfers_tree_A.transfers)
```

Comment:
So this uncovered a bug.
Here we should have `values_A.locksroot` and `values_A.locked`. At lines 151, 152 we should have `values_B.locksroot` and `values_B.locked`.

The problem is that tests pass because in the `TokenNetwork` smart contract, we have this mapping:
```
mapping (bytes32 => uint256) locksroot_to_locked_amount;
```
There is no way currently to see what participant provided what `locksroot`. Therefore, A can `unlock` its own merkle tree of pending transfers instead of the one provided by B, if this is more advantageous. This behavior is incorrect and leads to cheating. We need to fix this. Opened https://github.com/raiden-network/raiden-contracts/issues/157 for this.

So, for now, just do the swap between `values_A` and `values_B` and we will fix the bug later & add a test for it (with the logic that you actually used here).

Reply:
Yes, was thinking of adding the `participant_address` somewhere. E.g. `mapping (bytes32 => uint256) locksroot_to_locked_amount;` to be `keccak256(locksroot + participant_address) => locked_amount`.

But `keccak256(channel_identifier + participant_address) => UnlockData ` might also work.
I'm still thinking about what would be the best approach.

Reply:
The key point here is to know who the locked tokens belong to. If you link a `locksroot` to a `participant_address` in the smart contract, you know that. This will be done at `closeChannel` & `updateNonClosingBalance`, where you already have signatures.

E.g. `locksroot_A` is the merkle root of `A`, meaning that A sent those pending transfers to B -> you know who the `unlocked_amount` and `not_unlocked_amount` belongs to. 

So I don't see the reason for adding yet another pair of signatures for `unlock` if the smart contract already knows who has what `locksroot`. Am I missing something? 

In general, I think it would be optimal if:
1. we make sure tokens go where they belong (as explained above)
2. allow `unlock` to be called by anyone willing to pay the gas cost, as long as 1. is respected. This would allow people to use some additional systems/services where you pay for gas with tokens and the service makes the transaction for you.

Reply:
Using the signatures from `closeChannel` and `updateNonClosingBalance`  to determine the locksroot ownership does the job, but mapping `locksroot -> participant_address` is ambiguous. I'd also think about using a signature to determine that at least for now, as changing `unlock data` or the mapping might require a lot more refactoring.

Reply:
I would map `keccak256(locksroot + participant_address) => locked_amount` - what is ambiguous about this?

I still don't understand what would a signature achieve. You would still need to change the unlock data even with the signature. IOW the signature would just be an additional change on top of the data structure changes I mentioned - and I don't see the reason.

But, feel free to spec out some code here, so I can see exactly what you have in mind. It is possible I don't understand your idea entirely.

Reply:
As is it is now in the PR with `keccak256(channel_identifier, participant, locksroot)` it is clear.
Signature alone would't help. Yet might be good to have standalone authentication in `unlock`